packages/react/dialog/src/Dialog.tsx

```diff
@@ -296,28 +358,70 @@ function getState(open: boolean) {
   return open ? 'open' : 'closed';
 }
 
+const LabelWarningContext = React.createContext({
+  contentName: CONTENT_NAME,
+  titleName: TITLE_NAME,
+  docsSlug: 'dialog',
+});
+
+type LabelWarningProps = {
+  contentRef: React.RefObject<React.ElementRef<typeof DialogContent>>;
+};
+
+const LabelWarning: React.FC<LabelWarningProps> = ({ contentRef }) => {
+  const labelWarningContext = React.useContext(LabelWarningContext);
+
+  const MESSAGE = `\`${labelWarningContext.contentName}\` requires a label for the component to be accessible for screen reader users.
+
+You can label the \`${labelWarningContext.contentName}\` by passing a \`${labelWarningContext.titleName}\` component as a child, which also benefits sighted users by adding visible context to the dialog.
+
+Alternatively, you can use your own component as a title by assigning it an \`id\` and passing the same value to the \`aria-labelledby\` prop in \`${labelWarningContext.contentName}\`. If the label is confusing or duplicative for sighted users, you can also pass a label directly by using the \`aria-label\` prop.
+
+For more information, see https://radix-ui.com/primitives/docs/components/${labelWarningContext.docsSlug}`;
+
+  React.useEffect(() => {
+    const hasLabel =
+      contentRef.current?.getAttribute('aria-label') ||
+      document.getElementById(contentRef.current?.getAttribute('aria-labelledby')!);
+
+    if (!hasLabel) console.warn(MESSAGE);
+  }, [MESSAGE, contentRef]);
+
+  return null;
+};
+
 const Root = Dialog;
 const Trigger = DialogTrigger;
 const Overlay = DialogOverlay;
 const Content = DialogContent;
+const Title = DialogTitle;
+const Description = DialogDescription;
 const Close = DialogClose;
 
 export {
   Dialog,
   DialogTrigger,
   DialogOverlay,
   DialogContent,
+  DialogTitle,
+  DialogDescription,
   DialogClose,
   //
   Root,
   Trigger,
   Overlay,
   Content,
+  Title,
+  Description,
   Close,
+  //
+  LabelWarningContext,
```

Comment:
nit: Can we expose just `LabelWarningProvider` instead please? It looks like we only need the provider externally and we could also simplify consumption then:

```jsx
const LabelWarningProvider: React.FC<LabelWarningContextValue> = props => {
  return <LabelWarningContext.Provider value={props} />;
}
```

```jsx
<Dialog.LabelWarningProvider contentName="" titleName="" docsSlug="">
</Dialog.LabelWarningProvider>
``` 

vs

```jsx
<Dialog.LabelWarningContext.Provider value={React.useMemo(() => ({ contentName: '', titleName: '', docsSlug: '' ), [])}>
</Dialog.LabelWarningContext.Provider>
``` 

I'm not too bothered about doing the memoization internally though, it was more `Dialog.LabelWarningProvider` vs `Dialog.LabelWarningContext.Provider` that was bothering me so the following would suffice if you prefer 😅

```jsx
const LabelWarningProvider = LabelWarningContext.Provider;
export { LabelWarningProvider };
```

Reply:
I went with the latter, because the former was missing the fact it should render `children` which brings us back in the same situation as `createContext`.